vql/linux/audit/audit_service.go

```diff
@@ -0,0 +1,1018 @@
+}
+
+func (self *auditService) acceptEvents(ctx context.Context, sockFd int) error {
+	receivedCount := 0
+	discardedCount := 0
+	queuedCount := 0
+
+	// We're in non-blocking mode.  Try to get all of the events we can in one go.
+	var err error
+	for {
+		err = ctx.Err()
+		if err != nil {
+			break
+		}
+
+		buf := self.rawBufPool.Get().([]byte)
+		msgType, size, err := self.receiveMessageBuf(sockFd, buf)
+		if err != nil {
+			self.rawBufPool.Put(buf)
+			break
+		}
+
+		receivedCount += 1
+
+		// Messages from 1300-2999 are valid audit messages.
+		if msgType < auparse.AUDIT_USER_AUTH || msgType > auparse.AUDIT_LAST_USER_MSG2 {
+			self.rawBufPool.Put(buf)
+			discardedCount += 1
+			continue
+		}
+
+		self.messageQueue.Send(buf[:size])
+		queuedCount += 1
+	}
+
+	self.currentMessagesQueuedCounter.Add(queuedCount)
+	self.totalMessagesReceivedCounter.Add(receivedCount)
+	self.totalMessagesDiscardedCounter.Add(discardedCount)
+
+	if errors.Is(err, unix.EAGAIN) ||
+	   errors.Is(err, unix.EWOULDBLOCK) {
+		err = nil
+	}
+
+	return err
+}
+
+func (self *auditService) processOneMessage(buf []byte) error {
+	header := *(*unix.NlMsghdr)(unsafe.Pointer(&buf[0]))
+	msgType := auparse.AuditMessageType(header.Type)
+	data := buf[unix.NLMSG_HDRLEN:]
+
+	msgBuf := self.msgPool.Get().(*auparse.AuditMessage)
+	err := auparse.ParseBytes(msgType, data, msgBuf)
+	if err != nil {
+		self.msgPool.Put(msgBuf)
+		return err
+	}
+
+	self.reassembler.PushMessage(msgBuf)
+
+	// These record types aren't included in the complete callback
+	// but they still need to be pushed
+	if msgBuf.RecordType == auparse.AUDIT_EOE {
+		self.msgPool.Put(msgBuf)
+		return nil
+	}
+	self.totalOutstandingMessageCounter.Inc()
+	return nil
+}
+
+func (self *auditService) addSubscriberRules(subscriber *subscriber) error {
+	added := []*AuditRule{}
+
+	for _, rule := range subscriber.rules {
+		err := self.addRule(rule)
+		if err != nil {
+			// This will at minimum roll back the refcounts
+			for _, addedRule := range added {
+				self.deleteRule(addedRule)
+			}
+			return err
+		}
+		added = append(added, rule)
+	}
+
+	return nil
+}
+
+func (self *auditService) removeSubscriberRules(subscriber *subscriber) error {
+	for _, rule := range subscriber.rules {
+		err := self.deleteRule(rule)
+		if err != nil {
+			msg := fmt.Sprintf("audit: failed to remove rule `%s' during unsubscribe: %s", rule.rule, err)
+			subscriber.logChannel <- msg
+			continue
+		}
+	}
+
+	return nil
+}
+
+func (self *auditService) mainEventLoop(ctx context.Context) error {
+	self.Debug("audit: mainEventLoop started")
+	defer self.Debug("audit: mainEventLoop exited")
+	wg := sync.WaitGroup{}
+	defer wg.Wait()
+
+	for {
+		select {
+		case <- ctx.Done():
+			return ctx.Err()
+
+		case buf, ok := <- self.messageQueue.Output():
+			if !ok {
+				return nil
+			}
+			err := self.processOneMessage(buf)
+			if err != nil {
+				self.logger.Info("failed to parse message: %v", err)
+			}
+			self.rawBufPool.Put(buf)
```

Comment:
testing this branch with
```
velociraptor-v0.7.0-linux-amd64 -v query "select * from audit(rules=\"-a always,exit -F arch=b64 -S execve -k vrr_procmon\")" 
````
and running commands to trigger events, i'm finding most events are empty (fields are not filled in). The first event is always okay, but the next ones are likely to be empty. I think it's something to do with reusing buffers from the pool, because the events have all their fields filled if I comment out this line.

Reply:
Yep, that was definitely it. I now have test cases to ensure that the data output is sane.